### PR TITLE
Support complex numbers in DTensor redistribute

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157329

Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k

### CODE REVIEW

**Comment by wconstab** [comment]
@wconstab has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77564148).

File: test/distributed/tensor/test_redistribute.py [hunk 1]
```diff
@@ -116,6 +127,8 @@ def test_replicate_to_local_partial_grad(self):
             )
             out.backward(torch.ones_like(out))
 
+        # TODO(whc) it appears complex-allreduce is already being supported becuase this test passes,
+        # but I did not see where the support is
```
Comment (XilunWu) [inline test/distributed/tensor/test_redistribute.py]: this support is only for NCCL and some reduceOp: https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L4393-L4400

Comment (wconstab) [reply test/distributed/tensor/test_redistribute.py]: ah, thanks for pointing that out. That was the last piece i was missing. 

The new piece of info was that CI failed this test for GLOO bc of gloo not supporting allreduce.  I therefore added the complex support at Functional.cpp for allreduce, fixing the gloo test, and now I understand why nccl was already passing without it.  let me remove this TODO.

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 2]
```diff
@@ -81,8 +81,11 @@ at::Tensor all_reduce(
     const at::Tensor& input,
     std::string reduce_op,
     std::string group_name) {
-  auto output = input.clone(at::MemoryFormat::Contiguous);
-  return all_reduce_(output, std::move(reduce_op), std::move(group_name));
+  auto input_real = input.is_complex() ? at::view_as_real(input) : input;
+  auto output = input_real.clone(at::MemoryFormat::Contiguous);
+  auto output_ret =
+      all_reduce_(output, std::move(reduce_op), std::move(group_name));
+  return input.is_complex() ? at::view_as_complex(output_ret) : output_ret;
```
Comment (XilunWu) [inline torch/csrc/distributed/c10d/Functional.cpp:98]: We need the same logic (check reduce_op) (https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L4393-L4400) because this approach only holds numeric correctness for these 4 ops.

**Review by XilunWu — COMMENTED** [review]
The test change LGTM but the functional collective part needs small change. Thanks Will for adding complex number support!

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 2 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/Functional.cpp:98]: i don't know what the best choice is here to reuse the existing helper.  I ended up writing a new string-matching check.  Lmk if you think its worthwhile to do more refactoring to use the same helper from ProcessGroupNccl, but note that to use that i'd have to first convert into ReduceOp enum, and the conversion helper in this file is incomplete for some reason.  (premul_sum)

File: test/distributed/tensor/test_redistribute.py [hunk 3]
```diff
@@ -581,6 +600,9 @@ def test_shard_dim_alltoall(self):
         self.assertEqual(new_tensor.stride(), new_meta_tensor.stride())
 
 
+instantiate_parametrized_tests(RedistributeTest)
```
Comment (XilunWu) [inline test/distributed/tensor/test_redistribute.py:603]: note to myself: use `instantiate_parametrized_tests` in DTensor tests

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 4]
```diff
@@ -81,8 +81,21 @@ at::Tensor all_reduce(
     const at::Tensor& input,
     std::string reduce_op,
     std::string group_name) {
-  auto output = input.clone(at::MemoryFormat::Contiguous);
-  return all_reduce_(output, std::move(reduce_op), std::move(group_name));
+  if (input.is_complex()) {
+    TORCH_CHECK(
+        // TODO - ideally use 'to_reduce_op' helper but it currently errors on
+        // premul_sum
+        reduce_op == "sum" || reduce_op == "avg" || reduce_op == "premul_sum" ||
```
Comment (XilunWu) [inline torch/csrc/distributed/c10d/Functional.cpp:88]: IMO it would be good to reuse `complexViewAsRealAllowed`.

**Review by XilunWu — APPROVED** [review]
stamp to unblock. We can address the comment in follow-up PR

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 4 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/Functional.cpp:88]: i agree but it is not trivial.
1) convert string to enum: there is a helper in this file, but it is not complete. if i completed it by filling in missing premul_sum, i would be affecting the behavior of other ops using this helper
2) then i could refactor the complexViewAsRealAllowed out of ProcessGroupNccl.cpp and make it a util and use it.

Happy to do this in another PR, lmk if you have thoughts about (1)

Comment (XilunWu) [reply torch/csrc/distributed/c10d/Functional.cpp:88]: Low prio though. 

> lmk if you have thoughts about (1)

I think you're right. And I don't have a good solution either that can ensure consistency among helpers and potential extension of ReduceOp enum.
